doc/build-helpers/trivial-build-helpers.chapter.md

```diff
@@ -58,24 +58,115 @@ Variant of `runCommand` that forces the derivation to be built locally, it is no
 This sets [`allowSubstitutes` to `false`](https://nixos.org/nix/manual/#adv-attr-allowSubstitutes), so only use `runCommandLocal` if you are certain the user will always have a builder for the `system` of the derivation. This should be true for most trivial use cases (e.g., just copying some files to a different location or adding symlinks) because there the `system` is usually the same as `builtins.currentSystem`.
 :::
 
-## `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin` {#trivial-builder-writeText}
+## `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, `writeShellScriptBin` {#trivial-builder-textwriting}
 
-These functions write `text` to the Nix store. This is useful for creating scripts from Nix expressions. `writeTextFile` takes an attribute set and expects two arguments, `name` and `text`. `name` corresponds to the name used in the Nix store path. `text` will be the contents of the file. You can also set `executable` to true to make this file have the executable bit set.
+`nixpkgs` provides a number of functions that produce derivations which write text into the Nix store.  These include `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, and `writeShellScriptBin`, each of which is documented below.
 
-Many more commands wrap `writeTextFile` including `writeText`, `writeTextDir`, `writeScript`, and `writeScriptBin`. These are convenience functions over `writeTextFile`.
+These are useful for creating files from Nix expressions, which may be scripts or non-executable text files, depending on which of the functions is used and the arguments it takes.
+
+The result of each of these functions will be a derivation.  When you coerce the resulting derivation to text, it will evaluate to the *store path*. Importantly, it will not include the destination subpath produced by the particular function.  So, for example, given the following expression:
 
-Here are a few examples:
 ```nix
+
+my-file = writeTextFile {
+  name = "my-file";
+  text = ''
+    Contents of File
+  '';
+  destination = "/share/my-file";
+}
+```
+
+If `my-file` is coerced to text, it will resolve to `/nix/store/<store path>`, like any other derivation.  It will *not* evaluate to `/nix/store/<store path>/share/my-file`.  So to use it elsewhere, as an example (in this case, within a shell script you're writing in a Nix expression via `writeShellScript`), you might do:
+
+```nix
+writeShellScript "evaluate-my-file.sh" ''
+  cat ${my-file}/share/my-file
+'';
+```
+
+This will produce the desired result.  However, the following will not, it will fail because the store path is a directory, and is not the `my-file` file.
+
+```nix
+writeShellScript "evaluate-my-file.sh" ''
+  cat ${my-file}
+'';
+```
```

Comment:
I think this whole explanation can be shortened:
- There's no need to show examples that will fail. The general expectation when people see a code example in the manual is that it describes something that works. If you want to display an example of something that doesn't work, put it inside a warning admonition, but I think it's not needed here.
- There's no need to educate the reader that derivations can be coerced to store paths and the details on that. This is outside the scope of this section of the manual.

My suggestion is to reword this in a way that reads more like:
"Some of these functions will put the resulting files in a directory inside the derivation output. If you need to refer to the resulting files somewhere else in Nix code, remember to append the path to the file, like in the following example:
```
```nix
my-file = writeTextFile {
  name = "my-file";
  text = ''Contents of File';
  destination = "/share/my-file";
}

# Somewhere else in the code. Note how we had to append /share/my-file to the path!
writeShellScript "evaluate-my-file.sh" ''
  cat ${my-file}/share/my-file
'';
```

See [this tutorial](https://nix.dev/tutorials/nix-language#derivations) for more details on how a derivation is coerced into a store path."